File: backend/src/routes/scripts.ts
import fs from 'node:fs/promises';
import { Router } from 'express';
import { RUNS_DIR, snapshotFilePath, specFilePath } from '../config.js';
import { editSpec } from '../services/bedrock.js';
import { findRun, formatFailure } from '../services/failure.js';
import { GenerationError, generateAndSaveScript } from '../services/generator.js';
import { startRun } from '../services/runner.js';
import { runsStore, scriptsStore } from '../services/storage.js';

export const scriptsRouter: Router = Router();

scriptsRouter.post('/generate', async (req, res) => {
  const { url, prompt, name } = req.body ?? {};
  if (typeof url !== 'string' || !url.trim()) {
    return res.status(400).json({ error: 'A target URL is required.' });
  }
  if (typeof prompt !== 'string' || !prompt.trim()) {
    return res.status(400).json({ error: 'A test scenario description is required.' });
  }

  try {
    const result = await generateAndSaveScript({
      url: url.trim(),
      prompt: prompt.trim(),
      name: typeof name === 'string' ? name : undefined,
    });
    res.status(201).json(result);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    res.status(err instanceof GenerationError ? 502 : 500).json({ error: message });
  }
});

scriptsRouter.post('/run-all', async (_req, res) => {
  const scripts = await scriptsStore.read();
  if (scripts.length === 0) {
    return res.status(400).json({ error: 'There are no scripts to run.' });
  }

  // Queued in list order; the runner caps how many execute at once.
  const runs = [];
  for (const script of scripts) {
    runs.push(await startRun(script.id));
  }

  res.status(202).json(runs);
});

scriptsRouter.get('/', async (_req, res) => {
  res.json(await scriptsStore.read());
});

scriptsRouter.get('/:id', async (req, res) => {
  const script = (await scriptsStore.read()).find((s) => s.id === req.params.id);
  if (!script) return res.status(404).json({ error: 'Script not found.' });

  const code = await fs.readFile(specFilePath(script.id), 'utf-8').catch(() => '');
  res.json({ ...script, code });
});

scriptsRouter.put('/:id', async (req, res) => {
  const { name, code } = req.body ?? {};
  const script = (await scriptsStore.read()).find((s) => s.id === req.params.id);
  if (!script) return res.status(404).json({ error: 'Script not found.' });

  if (typeof code === 'string') {
    await fs.writeFile(specFilePath(script.id), code, 'utf-8');
  }

  const updated = await scriptsStore.update((scripts) =>
    scripts.map((s) =>
      s.id === script.id
        ? {
            ...s,
            name: typeof name === 'string' && name.trim() ? name.trim() : s.name,
            updatedAt: new Date().toISOString(),
          }
        : s,
    ),
  );

  res.json(updated.find((s) => s.id === script.id));
});

scriptsRouter.post('/:id/enhance', async (req, res) => {
  const { instruction, code, history, runId } = req.body ?? {};
  if (typeof instruction !== 'string' || !instruction.trim()) {
    return res.status(400).json({ error: 'Describe the change you want.' });
  }

  const script = (await scriptsStore.read()).find((s) => s.id === req.params.id);
  if (!script) return res.status(404).json({ error: 'Script not found.' });

  // Prefer the editor's current buffer over what's on disk — the user may have
  // unsaved edits they expect the assistant to build on.
  const current =
    typeof code === 'string' && code.trim()
      ? code
      : await fs.readFile(specFilePath(script.id), 'utf-8').catch(() => '');

  const snapshot = await fs
    .readFile(snapshotFilePath(script.id), 'utf-8')
    .catch(() => undefined);

  // Attached unprompted: when the last run is red, "fix the login step" should
  // reach the model already carrying the reason it went red. A specific runId
  // from the UI wins, so a user can point at an older failure.
  const run = await findRun(script.id, typeof runId === 'string' ? runId : undefined);
  const failure = run ? formatFailure(run) : undefined;

  try {
    const result = await editSpec({
      code: current,
      instruction: instruction.trim(),
      sourceUrl: script.sourceUrl,
      snapshot,
      failure,
      history: Array.isArray(history) ? history.slice(-8) : [],
    });
    res.json({ ...result, usedFailure: Boolean(failure) });
  } catch (err) {
    res.status(502).json({ error: err instanceof Error ? err.message : String(err) });
  }
});

scriptsRouter.delete('/:id', async (req, res) => {
  const { id } = req.params;
  const script = (await scriptsStore.read()).find((s) => s.id === id);
  if (!script) return res.status(404).json({ error: 'Script not found.' });

  await fs.rm(specFilePath(id), { force: true });
  await fs.rm(snapshotFilePath(id), { force: true });
  await fs.rm(`${RUNS_DIR}/${id}`, { recursive: true, force: true });
  await scriptsStore.update((scripts) => scripts.filter((s) => s.id !== id));
  await runsStore.update((runs) => runs.filter((r) => r.scriptId !== id));

  res.status(204).end();
});

scriptsRouter.get('/:id/runs', async (req, res) => {
  const runs = (await runsStore.read()).filter((r) => r.scriptId === req.params.id);
  res.json(runs);
});

scriptsRouter.post('/:id/runs', async (req, res) => {
  const script = (await scriptsStore.read()).find((s) => s.id === req.params.id);
  if (!script) return res.status(404).json({ error: 'Script not found.' });

  const run = await startRun(script.id);
  res.status(202).json(run);
});

File: backend-py/app/routers/scripts.py
from __future__ import annotations

import asyncio
import shutil
from typing import Any

from fastapi import APIRouter, Query, Request, Response

from ..config import RUNS_DIR, SCRIPTS_DIR, snapshot_file_path, spec_file_path
from ..http import ApiError, json_body, raw_text_field, text_field
from ..models import ScriptRecord
from ..services.bedrock import ChatTurn, edit_spec
from ..services.failure import find_run, format_failure
from ..services.runner import start_run
from ..services.steps import parse_steps
from ..services.storage import runs_store, scripts_store
from ..utils import now_iso

router = APIRouter()


async def _find_script(script_id: str) -> ScriptRecord:
    script = next((s for s in await scripts_store.read() if s.id == script_id), None)
    if not script:
        raise ApiError(404, "Script not found.")
    return script


def _read_text(path: Any) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except OSError:
        return ""


def _write_spec(script_id: str, code: str) -> None:
    # Saving an edit must not fail just because the scripts directory went missing.
    SCRIPTS_DIR.mkdir(parents=True, exist_ok=True)
    spec_file_path(script_id).write_text(code, encoding="utf-8")


@router.post("/run-all", status_code=202)
async def run_all(domain_id: str | None = Query(None, alias="domainId")) -> list[dict[str, Any]]:
    """Runs everything, or everything under one site when domainId is given."""
    scripts = await scripts_store.read()
    if domain_id:
        scripts = [script for script in scripts if script.domain_id == domain_id]
    if not scripts:
        raise ApiError(400, "There are no scripts to run.")

    # Queued in list order; the runner caps how many execute at once.
    runs = [await start_run(script.id) for script in scripts]
    return [run.dump() for run in runs]


@router.post("/steps")
async def preview_steps(request: Request) -> dict[str, Any]:
    """Plain-English reading of whatever code the editor currently holds.

    Takes the code rather than a script id so the steps view keeps up with
    unsaved edits — the two views should never disagree about what the test does.
    """
    code = raw_text_field(await json_body(request), "code") or ""
    return {"steps": [step.dump() for step in parse_steps(code)]}


@router.get("")
async def list_scripts(
    domain_id: str | None = Query(None, alias="domainId"),
) -> list[dict[str, Any]]:
    scripts = await scripts_store.read()
    if domain_id:
        scripts = [script for script in scripts if script.domain_id == domain_id]
    return [script.dump() for script in scripts]


@router.get("/{script_id}")
async def get_script(script_id: str) -> dict[str, Any]:
    script = await _find_script(script_id)
    code = await asyncio.to_thread(_read_text, spec_file_path(script.id))
    return {**script.dump(), "code": code, "steps": [step.dump() for step in parse_steps(code)]}


@router.put("/{script_id}")
async def update_script(script_id: str, request: Request) -> dict[str, Any] | None:
    body = await json_body(request)
    script = await _find_script(script_id)

    code = body.get("code")
    if isinstance(code, str):
        await asyncio.to_thread(_write_spec, script.id, code)

    name = text_field(body, "name")

    def mutate(scripts: list[ScriptRecord]) -> list[ScriptRecord]:
        return [
            s.model_copy(update={"name": name or s.name, "updated_at": now_iso()})
            if s.id == script.id
            else s
            for s in scripts
        ]

    updated = await scripts_store.update(mutate)
    found = next((s for s in updated if s.id == script.id), None)
    return found.dump() if found else None


@router.post("/{script_id}/enhance")
async def enhance(script_id: str, request: Request) -> dict[str, Any]:
    body = await json_body(request)
    instruction = text_field(body, "instruction")
    if not instruction:
        raise ApiError(400, "Describe the change you want.")

    script = await _find_script(script_id)

    # Prefer the editor's current buffer over what's on disk — the user may have
    # unsaved edits they expect the assistant to build on.
    current = raw_text_field(body, "code") or await asyncio.to_thread(
        _read_text, spec_file_path(script.id)
    )

    snapshot = await asyncio.to_thread(_read_text, snapshot_file_path(script.id))

    # Attached unprompted: when the last run is red, "fix the login step" should
    # reach the model already carrying the reason it went red. A specific runId
    # from the UI wins, so a user can point at an older failure.
    run_id = body.get("runId")
    run = await find_run(script.id, run_id if isinstance(run_id, str) else None)
    failure = format_failure(run) if run else None

    raw_history = body.get("history")
    history: list[ChatTurn] = [
        turn
        for turn in (raw_history if isinstance(raw_history, list) else [])
        if isinstance(turn, dict)
        and turn.get("role") in ("user", "assistant")
        and isinstance(turn.get("text"), str)
    ][-8:]

    try:
        result = await edit_spec(
            code=current,
            instruction=instruction,
            source_url=script.source_url,
            snapshot=snapshot or None,
            failure=failure,
            history=history,
        )
    except Exception as err:
        raise ApiError(502, str(err)) from err

    return {**result, "usedFailure": bool(failure)}


@router.delete("/{script_id}", status_code=204)
async def delete_script(script_id: str) -> Response:
    await _find_script(script_id)

    await asyncio.to_thread(_purge_files, script_id)
    await scripts_store.update(lambda scripts: [s for s in scripts if s.id != script_id])
    await runs_store.update(lambda runs: [r for r in runs if r.script_id != script_id])

    return Response(status_code=204)


def _purge_files(script_id: str) -> None:
    spec_file_path(script_id).unlink(missing_ok=True)
    snapshot_file_path(script_id).unlink(missing_ok=True)
    shutil.rmtree(RUNS_DIR / script_id, ignore_errors=True)


@router.get("/{script_id}/runs")
async def list_runs(script_id: str) -> list[dict[str, Any]]:
    return [run.dump() for run in await runs_store.read() if run.script_id == script_id]


@router.post("/{script_id}/runs", status_code=202)
async def create_run(script_id: str) -> dict[str, Any]:
    script = await _find_script(script_id)
    run = await start_run(script.id)
    return run.dump()
